python/rmm/rmm/_lib/memory_resource.pyx

```diff
@@ -229,6 +235,19 @@ cdef extern from "rmm/mr/device/failure_callback_resource_adaptor.hpp" \
             void* callback_arg
         ) except +
 
+cdef extern from "rmm/mr/device/failure_alternate_resource_adaptor.hpp" \
+        namespace "rmm::mr" nogil:
+    cdef cppclass failure_alternate_resource_adaptor[ExceptionType](
+        device_memory_resource
+    ):
+        # Notice, `failure_alternate_resource_adaptor` takes `device_async_resource_ref`
+        # as upstream arguments but we define them here as `device_memory_resource*` and
+        # rely on implicit type conversion.
+        failure_alternate_resource_adaptor(
+            device_memory_resource* upstream_mr,
+            device_memory_resource* alternate_upstream_mr,
+        ) except +
```

Comment:
Won't this break if the user actually passes a resource_ref rather than a pointer?

Reply:
No, since if the user did construct a `resource_ref` it would have a different type and the cython transpiler would complain during type-checking that it couldn't find the constructor.

But notice at the moment there's no way for the user to construct a `resource_ref` and pass it in cython.

Reply:
That isn't possible currently as we done have `device_async_resource_ref` bindings. But yes, if we later implement `device_async_resource_ref` bindings, we need to update the signature of `failure_alternate_resource_adaptor`.
I actually tried to write `device_async_resource_ref` bindings but it led to a range of issues.